smartdashboard/views.py

```diff
@@ -341,3 +343,448 @@ def __init__(
+
+    def update(self) -> None:
+        """Update all views within the OverviewView"""
+        self.exp_view.update()
+        self.app_view.update()
+        self.ens_view.update()
+        self.orc_view.update()
+
+
+class DualView(ViewBase):
+    """Base class for Telemetry Views that have table and chart elements"""
+
+    def __init__(
+        self,
+        shard: t.Optional[Shard],
+        table_element: DeltaGenerator,
+        graph_element: DeltaGenerator,
+        export_button: DeltaGenerator,
+    ):
+        self.shard = shard
+        self.table_element = table_element
+        self.graph_element = graph_element
+        self.export_button = export_button
+        self.telemetry = True
+        self.window_size = 10000
+        self.files = self._get_files()
+        self.message = self._get_message()
+        self.chart: t.Optional[alt.Chart] = None
+        self.columns = self.get_columns()
+        self.graph_delta_df = pd.DataFrame(columns=self.columns)
+
+        if self.shard is not None:
+            if self.files[0] != "" and self.files[1] != "":
+                try:
+                    self.telemetry_df = self._load_data()
+                except FileNotFoundError:
+                    self.table_element.info(self.message)
+                    self.telemetry = False
+            else:
+                self.table_element.info(self.message)
+                self.telemetry = False
+
+        else:
+            self.telemetry = False
+
+        if self.shard is not None and self.telemetry:
+            self.timestamp_min = self.telemetry_df["timestamp"].min()
+            self.initial_load()
+
+        self.enable_export_button()
+
+    @abstractmethod
+    def enable_export_button(self) -> None:
+        """Create an export data button"""
+
+    @abstractmethod
+    def _get_files(self) -> t.Tuple[str, str]:
+        """Returns a tuple of the files used for telemetry"""
+
+    @abstractmethod
+    def _get_message(self) -> str:
+        """Returns the error message if files cannot be found"""
+
+    @abstractmethod
+    def _handle_data(self) -> None:
+        """Updates the table and graph with appropriate dataframes"""
+
+    @abstractmethod
+    def initial_load(self) -> None:
+        """Updates the table and graph with initial dataframes"""
+
+    @abstractmethod
+    def get_columns(self) -> t.List[str]:
+        """Returns columns for the graph dataframe"""
+
+    def update(self) -> None:
+        """Checks for new data and calls to update the table
+        and graph if there is new data"""
+        if self.shard is not None and self.telemetry:
+            self.graph_delta_df = self._load_data_update(
+                skiprows=self.telemetry_df.shape[0]
+            )
+            if not self.graph_delta_df.empty:
+                self.telemetry_df = pd.concat(
+                    (self.telemetry_df, self.graph_delta_df), axis=0, ignore_index=True
+                )
+                self._handle_data()
```

Comment:
From what I can tell, this is the only place (outside of `__init__`) where `self.graph_delta_df` is assigned to, and the overloads of `self._handle_data` are the only places where `self.graph_delta_df` are read from.

From what I can tell the entire purpose of `self._graph_delta_df` is to act as input for `self._handle_data`. Is this true? If it is, can we pass the delta df as a parameter rather than making it an instance variable??

```suggestion
    def update(self) -> None:
        """Checks for new data and calls to update the table
        and graph if there is new data"""
        if self.shard is not None and self.telemetry:
            graph_delta_df = self._load_data_update(
                skiprows=self.telemetry_df.shape[0]
            )
            if not graph_delta_df.empty:
                self.telemetry_df = pd.concat(
                    (self.telemetry_df, graph_delta_df), axis=0, ignore_index=True
                )
                self._handle_data(graph_delta_df)
```
Do we even need to know the delta in the overloads of `self._handle_data`. It looks like they only use if there are not enough points in `self.telemetry_df` to warrant downsampling, in which case I'm not sure why we display the delta instead of all of `self.telemetry_df`.

Reply:
`graph_delta_df` is now a parameter passed in! And we do need to know the delta because if a `self.chart` is already created we only need to add in the the rows from `graph_delta_df` rather than regraph the entire dataframe.

`self.graph_element.add_rows(dframe)` can be used rather than 
`self.graph_element.altair_chart(
                self.chart, use_container_width=True, theme="streamlit"
            )`

Reply:
Gotcha! That makes sense to me!!